service/s3/s3crypto/kms_key_provider.go

```diff
@@ -0,0 +1,127 @@
+package s3crypto
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/service/kms"
+)
+
+// KMSKeyProvider will make calls to KMS to get the masterkey
+type KMSKeyProvider struct {
+	kms          *kms.KMS
+	key          []byte
+	encryptedKey []byte
+	iv           []byte
+	cmkID        *string
+
+	CipherData
+	MaterialDescription
```

Comment:
Change this to be aliased to map[string]string and keep marshaling/unmarshaling private